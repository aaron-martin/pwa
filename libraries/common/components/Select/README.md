# Select
---
The Select component is designed to be a custom select field for a given list of items.

## Getting Started

```markup
import { Select } from 'Library/components';

<Select
  items={['one', 'two', 'three']}
  onChange={console.log}
/>
```

## Props

### className

_Default_: `''`<br>
_Type_: `string`<br>

Additional classes that are to be appended to the Select container.

###### Usage:

```
<Select className="class1 class2"> ... </List>
```

### items

_Default_: `[]`<br>
_Type_: `Array`<br>

A list of items that can be selected.

Right now these item notations are supported:

```
[
  'item',
  { value: 'item value' },
  { value: 'item value', label: 'item label' },
]
```

*NOTE:* When you add an item as a string, this string will serve as it's value and label.


###### Usage:

```
<Select items={['one', 'two', 'three']} />
```

### value

_Default_: `undefined`<br>
_Type_: `string|number`<br>

The value of the item that should be selected initially.

###### Usage:

```
// This will select item 'two' initially
<Select items={['one', 'two', 'three']} value="two" />
```

### onChange

_Default_: `undefined`<br>
_Type_: `Function`<br>

A callback function that is called everytime the selection changed.
The callback takes the selected value as first and only parameter.

###### Usage:

```
<Select
  items={['one', 'two', 'three']}
  onChange={(selectedValue) => { console.log(selectedValue) }}
/>
```

### placeholder

_Default_: `'Select ...'`<br>
_Type_: `string`<br>

The default text for the component when no selection has been made.

###### Usage:

```
<Select placeholder="Please make a selection ..." />
```
