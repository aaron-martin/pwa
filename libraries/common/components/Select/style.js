/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

const container = css({
  margin: 0,
  padding: 0,
}).toString();

const selectHandle = css({
  float: 'right',
}).toString();

const items = css({
  position: 'absolute',
  width: '100%',
}).toString();

export default {
  container,
  selectHandle,
  items,
};
