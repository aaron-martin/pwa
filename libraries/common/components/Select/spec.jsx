/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Select from './index';
import SelectItem from './components/Item';

describe('<Select />', () => {
  jest.useFakeTimers();

  it('opens and closes the item list', () => {
    const wrapper = shallow(<Select />);

    let previousOpenState = wrapper.state('isOpen');
    wrapper.instance().toggleOpenState();
    expect(wrapper.state('isOpen')).toBe(!previousOpenState);

    previousOpenState = wrapper.state('isOpen');
    wrapper.instance().toggleOpenState();
    expect(wrapper.state('isOpen')).toBe(!previousOpenState);
  });

  it('renders without items', () => {
    const wrapper = mount(<Select />);

    wrapper.instance().toggleOpenState();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(SelectItem).length).toBe(0);
  });

  it('renders with implicit items (closed)', () => {
    const items = ['a', 'b', 'c', 'd', 'e', 'f'];

    const wrapper = mount(<Select items={items} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(SelectItem).length).toBe(0);
  });

  it('renders with implicit items (opened)', () => {
    const items = ['a', 'b', 'c', 'd', 'e', 'f'];

    const wrapper = mount(<Select items={items} />);
    wrapper.instance().toggleOpenState();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(SelectItem).length).toBe(items.length);
  });

  it('accepts implicit and explicit items', () => {
    const items = [
      'a',
      'b',
      {
        value: 'c',
      },
      'd',
      {
        value: 'e',
        label: 'E',
      },
      'f',
    ];

    const wrapper = mount(<Select items={items} />);

    wrapper.instance().toggleOpenState();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(SelectItem).length).toBe(items.length);

    let i = 0;
    wrapper.find(SelectItem).forEach((item) => {
      let expectedLabel = items[i];

      if (expectedLabel.label) {
        expectedLabel = expectedLabel.label;
      }
      if (expectedLabel.value) {
        expectedLabel = expectedLabel.value;
      }

      expect(item.prop('label')).toBe(expectedLabel);
      i += 1;
    });
  });

  it('triggers callback on change', () => {
    const items = ['a', 'b', 'c', 'd', 'e', 'f'];
    const selectionIndex = Math.floor(items.length / 2);

    /**
     * Mocked callback for the onSelect event
     * @param {string} value Mocked value
     */
    const callback = (value) => {
      expect(value).toBe(items[selectionIndex]);
    };

    const wrapper = mount((
      <Select
        items={items}
        onChange={callback}
      />
    ));

    wrapper.instance().toggleOpenState();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    const node = wrapper.find(SelectItem).at(selectionIndex);

    node.prop('onSelect')(node.prop('value'), node.prop('label'));
  });
});
