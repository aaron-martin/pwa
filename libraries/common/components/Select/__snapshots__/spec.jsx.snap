// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> accepts implicit and explicit items 1`] = `
<Select
  className=""
  items={
    Array [
      "a",
      "b",
      Object {
        "value": "c",
      },
      "d",
      Object {
        "label": "E",
        "value": "e",
      },
      "f",
    ]
  }
  onChange={[Function]}
  placeholder="Select ..."
  value={null}
>
  <div
    className="css-7c9av6 "
  >
    <div
      onTouchStart={[Function]}
    >
      <span
        className="undefined "
      >
        Select ...
      </span>
      <span
        className="css-jf1dde"
      >
        ▾
      </span>
    </div>
    <div
      className="css-182i7wq"
    >
      <SelectItem
        className=""
        label="a"
        onSelect={[Function]}
        selected={null}
        value="a"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          a
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="b"
        onSelect={[Function]}
        selected={null}
        value="b"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          b
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="c"
        onSelect={[Function]}
        selected={null}
        value="c"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          c
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="d"
        onSelect={[Function]}
        selected={null}
        value="d"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          d
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="E"
        onSelect={[Function]}
        selected={null}
        value="e"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          E
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="f"
        onSelect={[Function]}
        selected={null}
        value="f"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          f
        </div>
      </SelectItem>
    </div>
  </div>
</Select>
`;

exports[`<Select /> renders with implicit items (closed) 1`] = `
<Select
  className=""
  items={
    Array [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
    ]
  }
  onChange={[Function]}
  placeholder="Select ..."
  value={null}
>
  <div
    className="css-7c9av6 "
  >
    <div
      onTouchStart={[Function]}
    >
      <span
        className="undefined "
      >
        Select ...
      </span>
      <span
        className="css-jf1dde"
      >
        ▾
      </span>
    </div>
  </div>
</Select>
`;

exports[`<Select /> renders with implicit items (opened) 1`] = `
<Select
  className=""
  items={
    Array [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
    ]
  }
  onChange={[Function]}
  placeholder="Select ..."
  value={null}
>
  <div
    className="css-7c9av6 "
  >
    <div
      onTouchStart={[Function]}
    >
      <span
        className="undefined "
      >
        Select ...
      </span>
      <span
        className="css-jf1dde"
      >
        ▾
      </span>
    </div>
    <div
      className="css-182i7wq"
    >
      <SelectItem
        className=""
        label="a"
        onSelect={[Function]}
        selected={null}
        value="a"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          a
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="b"
        onSelect={[Function]}
        selected={null}
        value="b"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          b
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="c"
        onSelect={[Function]}
        selected={null}
        value="c"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          c
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="d"
        onSelect={[Function]}
        selected={null}
        value="d"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          d
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="e"
        onSelect={[Function]}
        selected={null}
        value="e"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          e
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="f"
        onSelect={[Function]}
        selected={null}
        value="f"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          f
        </div>
      </SelectItem>
    </div>
  </div>
</Select>
`;

exports[`<Select /> renders without items 1`] = `
<Select
  className=""
  items={Array []}
  onChange={[Function]}
  placeholder="Select ..."
  value={null}
>
  <div
    className="css-7c9av6 "
  >
    <div
      onTouchStart={[Function]}
    >
      <span
        className="undefined "
      >
        Select ...
      </span>
      <span
        className="css-jf1dde"
      >
        ▾
      </span>
    </div>
    <div
      className="css-182i7wq"
    />
  </div>
</Select>
`;

exports[`<Select /> triggers callback on change 1`] = `
<Select
  className=""
  items={
    Array [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
    ]
  }
  onChange={[Function]}
  placeholder="Select ..."
  value={null}
>
  <div
    className="css-7c9av6 "
  >
    <div
      onTouchStart={[Function]}
    >
      <span
        className="undefined "
      >
        Select ...
      </span>
      <span
        className="css-jf1dde"
      >
        ▾
      </span>
    </div>
    <div
      className="css-182i7wq"
    >
      <SelectItem
        className=""
        label="a"
        onSelect={[Function]}
        selected={null}
        value="a"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          a
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="b"
        onSelect={[Function]}
        selected={null}
        value="b"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          b
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="c"
        onSelect={[Function]}
        selected={null}
        value="c"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          c
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="d"
        onSelect={[Function]}
        selected={null}
        value="d"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          d
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="e"
        onSelect={[Function]}
        selected={null}
        value="e"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          e
        </div>
      </SelectItem>
      <SelectItem
        className=""
        label="f"
        onSelect={[Function]}
        selected={null}
        value="f"
      >
        <div
          className="css-fisw11 "
          onTouchEnd={[Function]}
        >
          f
        </div>
      </SelectItem>
    </div>
  </div>
</Select>
`;
